optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,163 @@
+        self._resource_budget = 0
+        self._n_brackets = n_brackets
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hyperband has {} brackets'.format(self._n_brackets))
+
+        for i in range(n_brackets):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_brackets)
+            self._resource_budget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> bool
+
+        i = self._get_bracket_id(study, trial)
+        _logger.debug('{}th bracket is selected'.format(i))
+        bracket_study = self._create_bracket_study(study, i)
+        return self._pruners[i].prune(bracket_study, trial)
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_brackets):
+        # type: (int, int) -> int
+
+        n = self._reduction_factor ** (n_brackets - 1)
+        budget = n
+        for i in range(pruner_index, n_brackets - 1):
+            budget += n / 2
+        return budget
+
+    def _get_bracket_id(self, study, trial):
+        # type: (Study, structs.FrozenTrial) -> int
+        """Computes the index of bracket for a trial of ``trial_number``.
+
+        The index of a bracket is noted as :math:`s` in
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        n = hash('{}_{}'.format(study.study_name, trial.number)) % self._resource_budget
+        for i in range(self._n_brackets):
+            n -= self._bracket_resource_budgets[i]
+            if n < 0:
+                return i
+
+        assert False, 'This line should be unreachable.'
+
+    def _create_bracket_study(self, study, bracket_index):
+        # type: (Study, int) -> Study
+
+        # NOTE(crcrpar): The below import is workaround.
+        # The below import violates PEP8 (https://www.python.org/dev/peps/pep-0008/#imports),
+        # however, if we follow that here, ImportError occurs.
+
+        from optuna.study import Study
+
+        # N.B. This class is assumed to be passed to
+        # `SuccessiveHalvingPruner.prune` in which `get_trials`,
+        # `direction`, and `storage` are used.
+        # But for safety, prohibit the other attributes explicitly.
+        class _BracketStudy(Study):
```

Comment:
More of a question but is there any reason for making this one local (i.e. `...<locals>. _BracketStudy `)? I assume you'd generally define it a a (sub)module level.

Reply:
I don't understand the point. Is this related to the import of `optuna.study.Study` discussed above?

ref: https://github.com/optuna/optuna/pull/809/files#r363220739

Reply:
They're orthogonal. Replied in https://github.com/optuna/optuna/pull/809#discussion_r363561471 though.